### ISSUE TITLE
Cursor changes position between consecutive calls to complete functions

### ISSUE THREAD

**Issue by echasnovski** [vim/vim#8510]
**Describe the bug**
First and second calls to any of 'complete-function's (both omnifunc and completefunc) might have different cursor positions.

**To Reproduce**
Detailed steps to reproduce the behavior:
1. Save the following configuration file (for example, under 'completefunc-curpos.vim'):
```vim
let g:curpos_dict = {}
function g:CompleteFunc(findstart, base)
  let g:curpos_dict[a:findstart] = getpos('.')

  if a:findstart == 1
    return 2
  else
    return ['Jun', 'Jul']
  endif
endfunction

set omnifunc=g:CompleteFunc
```
2. Run `vim -u completefunc-curpos.vim` (or insert other name for configuration file)
3. In Insert mode type 'abcdefg'.
4. Type '\<C-x>\<C-o>' to call omnifunc.
5. Type '\<Esc>:echo g:curpos_dict\<Cr>'. This should show `{'0': [0, 1, 3, 0], '1': [0, 1, 8, 0]}` meaning that cursor changed position from [0, 1, 8, 0] to [0, 1, 3, 0] (start of completion item) between calls.

**Expected behavior**
Cursor position doesn't change from first to second call to 'complete-function'. This issue is even more apparent, when using `completeopt=noselect` as cursor doesn't visibly change its position.

**Screenshots**
Animation of behavior:

![vim_complete-function-curpos-change](https://user-images.githubusercontent.com/24854248/124389789-6d112000-dcf1-11eb-83f1-bfb0c9cb831c.gif)

**Environment (please complete the following information):**
 - Vim version [e.g. 8.1.1234] (Or paste the result of `vim --version`.)
 - OS: Ubuntu 18.04
 - Terminal: GNOME Terminal

**Additional context**
Cursor will not actually move after first call if it returned `-1` because it "moved" to the same spot.
Change in cursor position is relevant to writing custom completion functions.

**Comment by chrisbra** [comment]
I believe what happens is this: On the first call of the omnicompletion function, you tell vim to start completing from column 2. So Vim removes the old text until column 2, meaning the cursor cannot be at at position 8 anymore, since the line is too short. That's why on the second call the cursor position is at 2. Because that is where you want the completion to be inserted. Perhaps we should just document that the cursor will be positioned on the column that is returned on the first call of the completion function.

**Comment by echasnovski** [comment]
Yes, this seems like a the most reasonable explanation of the current behavior. And at least documentation would be nice, certainly (it would have saved some time of my life :) ).

I would argue, though, that this behavior is not really intuitive. I don't see a reason why active buffer should be modified at all when computing completion matches on the second call. Especially, when having `noselect` in `compleopt`: currently on second call Vim basically removes all text subject to completion, computes matches, and then pastes back previous text to achieve the exact starting state.
